PR: Spyder Workshops for research and finance (#261)
spyder-ide/spyder-docs: doc/workshops/financial.rst

* Basic knowledge of Statistics
* Some knowledge of Financial Econometrics is desirable

=====
Intro
=====

In this workshop we will obtain financial data in real-time and explore financial portfolios using Econometrics and computational tools.

Why use Python for financial analysis?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Real-time analysis of historical and current financial data is essential for those investing in financial instruments. Real-time analysis of historical and current financial data is essential for those investing in financial instruments. Python has a number of features that make it ideal for financial tasks:

* It is easy to learn for anyone, whether they have previous programming experience or not
* It has a variety of specialized mathematical and statistical libraries (SciPy, NumPy, Pandas) that are commonly used for financial analysis
* It can be connected to APIs for loading financial data in real time
* It is the programming language with the most resources for machine learning (supervised learning, unsupervised learning, reinforcement learning), which is one of the most useful tools for Econometrics nowadays
* It has excellent libraries for plotting

> * You can use resources such as Google Colab or Binder to do your analysis in the cloud.

Reviewer: I suggest adding links to these tools.

.. image:: images/workshop-1/set-up-environment.png
   :alt: Set up environment in Spyder's Preferences

Now, you have everything ready to proceed with the workshop.
   
   
   
=====================
Obtain financial data
=====================
   
When it comes to finance, being up to date is very important. So we are going to use a Python library that allows us to get updated historical Stock Market records from `Yahoo! Finance`_ API. In this way, we will be able to download data in the period of time we are interested in analyzing.

.. _`Yahoo! Finance`: https://finance.yahoo.com

To get started, import the libraries.

.. important:: 

   Write and run all code for this workshop in the "IPython Console" at the bottom left of Spyder.

.. code-block:: python
      
   import numpy as np
   import pandas as pd
   from pylab import mpl
   from scipy.optimize import minimize
   import matplotlib.pyplot as plt
   import pprint
   import yfinance as yf

> The first group of imports is for basic operations. The second imports the libraries that we will use to download financial data.

Reviewer: I don't know if it's useful to mention two groups, if they are actually merged into one in the code. Maybe leave a blank like between groups so this is clearer? Otherwise you can just mention the libraries in each group by name.

General stock information
~~~~~~~~~~~~~~~~~~~~~~~~~

If you want to know which methods and attributes you can query, you can do so with the built-in ``help()`` function.

You can also type the name of the object (``netflix``)in the console, then type a period and hit the Tab key once. IPython suggestions will then appear to help you navigate the object:

.. image:: images/workshop-1/ipython-auto-suggestions.png
   :alt: IPython Console auto-suggestions

For example, we can obtain general information with the ``info`` property of the object.

.. code-block:: python

   netflix_info = netflix.info

We can observe the result in the Variable Explorer, in the ``netflix_info`` variable. 

.. image:: images/workshop-1/net_flix-variable-in-variable-explorer.png
   :alt: First variable in Variable Explorer

If we double-click it, a new window with the detailed information will be displayed. 

.. image:: images/workshop-1/dict-in-variable-explorer.png
   :alt: Python dict example in Variable Explorer

> In that window we can see that a Python dictionary is displayed in which each key has a value assigned to it. Each type of value is represented by a distinctive color. For example, we see that Netflix is an Entertainment industry, the summary of the business type, among other indicators. The Variable Explorer is a very convenient way to view these types of results. We can compare it to viewing them in the Ipython Console with the ``pprint`` function:

Reviewer: ```suggestion
In that window we can see that a Python dictionary is displayed in which each key has a value assigned to it. Each type of value is represented by a distinctive color. For example, we see that Netflix is an Entertainment industry, the summary of the business type, among other indicators. The Variable Explorer is a very convenient way to view these types of results. We can compare it to viewing them in the IPython Console with the ``pprint`` function:
```

.. code-block:: python

   pprint.pprint(netflix_info)

Although ``pprint`` displays all the information, it is easier to view it in the Variable Explorer.

Historical stock data
~~~~~~~~~~~~~~~~~~~~~

We can download in a dataset the history of a stock with the following line of code:

.. code-block:: python

   hist = netflix.history(period="max")

We can see a summary and more details of this dataset in the Variable Explorer. It appears as a DataFrame object, with 7 columns and thousands of rows. If we double-click on it we will see the historical records of Netflix stock organized by ascending dates.

.. image:: images/workshop-1/netflix-history.png
   :alt: Netflix historical operations

Throughout this workshop we will use this historical information to compare, with various normative financial theories, different types of financial portfolios.

===============
First portfolio
===============

Let's build our first stock investment portfolio! Say we are interested in investing in technology, and we want to know what performance can be obtained by putting money into some of the "heavyweights" in this industry.

> To measure the performance of our first portfolio we are going to use a classic theory in the world of finance: *mean-variance portfolio (MVP) theory*. This model assumes that investors only care about expected returns and the variance of such returns. The analysis is based entirely on statistical measures based on a time series of share prices, such as periodic mean returns and the variances of those returns with the same periodicity.

Reviewer: It would be nice to have a link to some basic explanation of this theory (maybe Wikipedia)

.. code-block:: python

   monthly_data_1 = close_data_1.resample("M").ffill().pct_change()

Since each stock is a column, we can use the ``mean()`` method on the DataFrame to see the results.

.. code-block:: python

   monthly_data_1.mean()

   # AMZN    0.029931
   # GOOG    0.018784
   # MSFT    0.020698
   # AAPL    0.023033
   # NFLX    0.042060
   # dtype: float64

.. important::

   Remember that this closing information represents the relative growth of each stock, not its closing price in any currency.

As we can see, all the results are positive, so this portfolio has been profitable over the years. In relative terms, the biggest gains would come from Netflix (0.4 percent). These percentages show an average of the monthly variation of the stock price. But how stable have that prices been? To find this out, we can calculate the standard deviation of stock price growth:

.. code-block:: python

   monthly_data_1.std()

   # AMZN    0.082308
   # GOOG    0.061057
   # MSFT    0.058395
   # AAPL    0.081239
   # NFLX    0.145416
   # dtype: float64

> The highest volatility is found in Netflix (0.1454), which indicates that its price, despite being the fastest growing, has also had a lot of variation during these years. The most stable price, on the other hand, has been Microsoft's one (0.0583).

Reviewer: ```suggestion
The highest volatility is found in Netflix (0.1454), which indicates that its price, despite being the fastest growing, has also had a lot of variation during these years. The most stable price, on the other hand, has been Microsoft's (0.0583).
```

.. code-block:: python

   (close_data_1/ close_data_1.iloc[0]).plot(figsize=(16, 10), title="Portfolio 1 daily stock price")

.. image:: images/workshop-1/portfolio-1-daily-stock-price.png
   :alt: Portfolio 1 daily stock price

.. important::

   You will be able to see the graph in the Plots pane. On the left you will see the plot in detail. On the right, a stack will be created with all the plots that are generated both in the Console and by running the code directly in the Editor. You can copy, delete or save the plot to disk by clicking on it with the right mouse button or trackpad.

The above time-series plot clearly shows the growth of all stocks, particularly Netflix. The high volatility discussed above can also be seen. For example, it can be seen from the plots that, due to high volatility, investing in Netflix in the years 2014 and 2019 left virtually no profit (if shares were bought at the beginning of the year and sold right at the end).

> We can also plot with the ``plot()`` method the dataframe we already have with the monthly data. To do this, we will add 1 to all the values and calculate the accumulated product with the ``cumprod()`` method:

Reviewer: For consistency:
```suggestion
We can also plot with the ``plot()`` method the DataFrame we already have with the monthly data. To do this, we will add 1 to all the values and calculate the accumulated product with the ``cumprod()`` method:
```

.. note::

   A stock is a financial security that represents that you own a part of a company (whatever the size of this part). A share, on the other hand, is the smallest unit of denomination of a stock. A stock is composed of one or several shares.

This weight will be a vector (Python list) composed of the relative weight (between 0 and 1) of each stock in the portfolio. The sum of these weights has to be 1. Since we will assume that each stock is a single share, the distribution will be equal: ``[0.2, 0.2, 0.2, 0.2, 0.2]``.

.. code-block:: python

   weights_1 = [0.2, 0.2, 0.2, 0.2, 0.2]

With this information we can calculate the expected return of this portfolio. The math is simple: this is given by the dot product of the portfolio weights vector and the vector of expected returns. This result must be multiplied by the number of days for which the return is to be calculated (a year has approximately 252 stock price closings).

Let's put all this into a function:

.. code-block:: python

   def portfolio_return(returns, weights):
       return np.dot(returns.mean(), weights) * 252

Let's look at the expected return of this portfolio if we hold it for one year:

> .. code-block: python

Reviewer: ```suggestion
.. code-block:: python
```

This seems to be a good portfolio because it has a good performance with a not very large variance.

Optimal portfolio weights
~~~~~~~~~~~~~~~~~~~~~~~~~

Can we use the data obtained to calculate the optimal weights for the portfolio by year? Of course we can. Let's start by delimiting the previous years as variables.

.. code-block:: python

   start_year, end_year = (2012, 2020)

We will now write a function to calculate these optimal weights.

.. code-block:: python

   def optimal_weights(returns, symbols, actual_weights, start_y, end_y):
      
       bounds = len(symbols) * [(0, 1), ]
       constrains = {"type": "eq", "fun": lambda weights: weights.sum() - 1}
       opt_weights = {}
 
       for year in range(start_y, end_y):
           _rets = returns[symbols].loc[f"{year}-01-01":f"{year}-12-31"]
           _opt_w = minimize(lambda weights: -portfolio_sharpe(_rets, weights),
                          actual_weights,
                          bounds=bounds,
                          constraints=constrains)["x"]
           opt_weights[year] = _opt_w
       return opt_weights
       

> Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constrains`` variables). Notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.

Reviewer: Do you mean to use `constrains` instead of `constraints`? If yes, no problem, but it may be a source of confusion/errors for users if they expect the usual spelling.

Reviewer: ```suggestion
Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constrains`` variables). Notice the minus sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of ``portfolio_sharpe`` a negative one.
```

Reviewer 2: you missed this one and the previous comment as well

> We will use the function we just defined to calculate the optimal weights for each year. And we are going to save the result in a Pandas DataFrame to take advantage of the Variable Explorer display options.

Reviewer: ```suggestion
We will use the function we just defined to calculate the optimal weights for each year, and we are going to save the result in a Pandas DataFrame to take advantage of the Variable Explorer display options.
```

Let's summarize these numbers.

.. code-block:: python

   port_1_exp_real.mean()

   # expected    0.408009
   # realized    0.199700
   # dtype: float64

Our optimal weight model offered us a profit of around 40%, but the profit we would have obtained, due to real market fluctuations, would have been almost 20%. Not bad, but the mean-variance portfolio model we have applied for annual calculations is not very accurate, is it?

And this result is less encouraging if we calculate the correlations between expected and realized profits.

.. code-block:: python

   port_1_exp_real[["expected", "realized"]].corr()

   #           expected  realized
   # expected  1.000000 -0.324053
   # realized -0.324053  1.000000

As we can see, the correlations are negative, which warns us that we should be cautious when using this type of modeling.

================
Second portfolio
================

> We are now going to apply all the previous code with a portfolio of a different nature. Let's assume that instead of technology companies, we are now interested in pharmaceuticals. We will build a portfolio with stocks of Pfizer, Astra Zeneca, Johnson N Johnson.

Reviewer: I believe this should read
```suggestion
We are now going to apply all the previous code with a portfolio of a different nature. Let's assume that instead of technology companies, we are now interested in pharmaceuticals. We will build a portfolio with stocks of Pfizer, Astra Zeneca, Johnson & Johnson.
```

Download the data
~~~~~~~~~~~~~~~~~

> Let's download the data and format them.

Reviewer: I think it should be:

```suggestion
Let's download the data and format it.
```

   fig = plt.figure()
   (monthly_data_2 + 1).cumprod().plot(figsize=(16, 10), title="Portfolio 2 monthly stock price")

.. image:: images/workshop-1/portfolio-2-daily-stock-price.png
   :alt: Portfolio 2 daily stock price

.. image:: images/workshop-1/portfolio-2-monthly-stock-price.png
   :alt: Portfolio 2 monthly stock price

These two plots show a steady growth over the years, but also a high variability in each year. This seems to be a good portfolio only if taken as a long-term investment.

Returns, volatility and Sharpe ratio
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To confirm what was said in the previous section, let's calculate returns, volatility and Sharpe ratio.

.. code-block:: python

   weights_2 = len(close_data_2.columns) * [1 / len(close_data_2.columns)]

   print(f"Portfolio 2 returns: {portfolio_return(rets_2, weights_2):.4f}")
   print(f"Portfolio 2 volatility: {portfolio_volatility(rets_2, weights_2):.4f}")
   print(f"portfolio 2 Sharpe: {portfolio_sharpe(rets_2, weights_2):.4f}")

   # Portfolio 2 returns: 0.0809
   # Portfolio 2 volatility: 0.1637
   # Portfolio 2 Sharpe: 0.4940

> The return of this portfolio is significantly lower than that of the previous portfolio (0.0809 < 0.2859). Its volatility (0.1637 < 0.2370) is also lower, but to a lesser extent.This is reflected in a lower Sharpe ratio as well (0.4940 < 1.2062). This means, within the mean-variance theory approach, that the first portfolio is a better investment than the second one.

Reviewer: ```suggestion
The return of this portfolio is significantly lower than that of the previous portfolio (0.0809 < 0.2859). Its volatility (0.1637 < 0.2370) is also lower, but to a lesser extent. This is reflected in a lower Sharpe ratio as well (0.4940 < 1.2062). This means, within the mean-variance theory approach, that the first portfolio is a better investment than the second one.
```

The different behavior is clearly observed if we apply a Monte Carlo simulation and visualize it with a graph:

.. code-block:: python

   port_2_vr, port_2_sr = monte_carlo_sharpe(rets_2, SYMBOLS_2, weights_2)

   plt.figure(figsize=(16, 10))
   fig = plt.scatter(port_2_vr[:, 0], port_2_vr[:, 1], c=port_2_sr, cmap="cool")
   CB = plt.colorbar(fig)
   CB.set_label("Sharpe ratio")
   plt.xlabel("expected volatility")
   plt.ylabel("expected return")
   plt.title(" | ".join(SYMBOLS_2))

.. image:: images/workshop-1/scatter-plot-sharpe-portfolio-2.png
   :alt: Portfolio 2 Monte Carlo Sharpe simulation scatter plot

High volatility does not correspond in most cases with high returns. In fact, there are scenarios in the simulation in which higher expected returns are related to lower expected volatility.

> Optima weights

Reviewer: ```suggestion
Optimal weights
```

   print("Expected and realized means:")
   print(port_2_exp_real.mean())
   print("Expected and realized correlations:")
   print(port_2_exp_real[["expected", "realized"]].corr())

   # Expected and realized means:
   # expected    0.155103
   # realized    0.062207

   # Expected and realized correlations:
   #          expected  realized
   # expected  1.000000 -0.023134
   # realized -0.023134  1.000000

As we can see, the model with optimal weights predicted a return close to 15% per year, but the realized return would have barely reached 6%. And the negative correlations show, as with portfolio 1, that there does not seem to be any correspondence between these values.

The comparison is then favorable for portfolio 1. But what if we compare portfolio 1 with a "higher risk" investment such as cryptocurrencies? Let's discuss it below.

===============
Third portfolio
===============

Download cryptocurrencies data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Our third portfolio will consist of three cryptocurrencies: bitcoin (BTC), ethereum (ETH) and litecoin (LTC). To access historical data, we are going to use a library called **Historic-Crypto**.

.. important::

>    The Python Historic-Crypto library is not found in the Anaconda repository. It can be found in the PyPI repository and can be downloaded with the Pip package manager as follows: ``>>> pip install Historic-Crypto``. If you do not want to install this library, download the dataset "crypto_hist.csv" in your working directory, and load it in memory with the instruction ``crypto_hist = pd.read_csv("crypto_hist.csv")``, and skip the rest of this section.

Reviewer: Does this mean we expect people to use pip from inside the ipython terminal? Just to confirm.

Reviewer 2: This instruction was at the beginning of the workshop, why put it here?

Author: Fixed.

>    The Python Historic-Crypto library is not found in the Anaconda repository. It can be found in the PyPI repository and can be downloaded with the Pip package manager as follows: ``>>> pip install Historic-Crypto``. If you do not want to install this library, download the dataset "crypto_hist.csv" in your working directory, and load it in memory with the instruction ``crypto_hist = pd.read_csv("crypto_hist.csv")``, and skip the rest of this section.

Reviewer: Oh and maybe there should be a link to the csv file here.

Import the libraries:

.. code-block:: python

   from Historic_Crypto import Cryptocurrencies
   from Historic_Crypto import HistoricalData

We are going to use the ``Cryptocurrencies`` class to obtain a list of available cryptocurrencies.

.. code-block:: python

   crypto_list = Cryptocurrencies(coin_search="", extended_output=True).find_crypto_pairs()

.. note::

>    If you want more information about the use of this library you can make a quick query using the Spyder Help panel (type in the console ``Cryptocurrencies`` and use ctrl-I or cmd-I to display it).Or you can read the documentation in their `official repository <https://github.com/David-Woroniuk/Historic_Crypto>`_

Reviewer: ```suggestion
   If you want more information about the use of this library you can make a quick query using the Spyder Help panel (type in the console ``Cryptocurrencies`` and use ctrl-I or cmd-I to display it). Or you can read the documentation in their `official repository <https://github.com/David-Woroniuk/Historic_Crypto>`_
```

- Return: 0.6203
- Volatility: 0.7587
- Sharpe: 0.8176

These numbers are higher than those obtained for portfolios 1 and 2, except for the Sharpe ratio. This is a very volatile portfolio (in fact three times more volatile than that of the technology companies), and that makes it ultimately not as profitable as portfolio 1. The returns are higher (almost double) but the risk may not compensate for it.

Monte carlo simulation
~~~~~~~~~~~~~~~~~~~~~~

The Monte Carlo simulation also shows the non-linear correlation between risk and returns (as you can see, sometimes high risk involves only modest profits):

.. image:: images/workshop-1/scatter-plot-sharpe-portfolio-3.png
   :alt: Portfolio 3 Monte Carlo Sharpe simulation scatter plot

As can be seen, there are points (bottom right) that show a very high volatility and yet have a very low expected return. In this sense, portfolio 1 represents a safer investment because the higher risk is consistently offset by higher returns.

Optimal weights
~~~~~~~~~~~~~~~

.. image:: images/workshop-1/portfolio-3-optimal-weights.png
   :alt: Portfolio 3 optimal weights by year

> The optimal portfolio weights, if calculated annually, suggest that our portfolio should have been quite polarized in some years: the recommendation is to have bought only bitcoins before the start of 2016 and 2019, and only ethereum in 2018. Starting 2017 and 2020, on the other hand, our weights recommended a more balanced investment between bitcoin and ethereum. Litecoin is not largely recommended by our model.

Reviewer: ```suggestion
The optimal portfolio weights, if calculated annually, suggest that our portfolio should have been quite polarized in some years: the recommendation is to have bought only bitcoins before the start of 2016 and 2019, and only ethereum in 2018. Starting 2017 and 2020, on the other hand, our weights recommended a more balanced investment between bitcoin and ethereum. Litecoin is not recommended by our model.
```

Financial Data Analysis with Spyder
###################################

By the end of this workshop participants will be able to use Spyder effectively for  applying some normative financial theories and models to assemble a portfolio of assets in a way that maximizes the expected return for a given level of risk. In this way, statistically-based tools are used to construct investment portfolios.

=============
Prerequisites
=============

To follow this workshop we recommend that you have a basic knowledge of Python. You can visit `The Python Tutorial`_ to learn the basics of this programming language or to refresh your knowledge of Python.

.. _The Python Tutorial: https://docs.python.org/3/tutorial/

> You will also need to have Spyder installed. More information about installation in :doc:`installation guide<../installation>`.

Reviewer 2: I think they need Anaconda since they need them to create environments and install packages

   
   $ conda env create -f workshop-1-environment.yml

You can now enter the newly created virtual environment in this way:

.. code-block:: bash
   
   $ conda activate financial-analysis

   
Download the datasets
~~~~~~~~~~~~~~~~~~~~~
   
Although during the workshop we will explain how to use some APIs to download up-to-date data, you can also download the datasets in csv format from `this link <https://figshare.com/articles/dataset/Historical_financial_datasets_for_Financial_Analysis_with_Spyder_workshop/14995215>`_.

To follow this workshop you do not need to create a new directory. However, if you have downloaded the data and want to use it instead of the APIs, you must set the directory that has the downloaded data as the working directory. In order to do this, check that the working directory is correct. You should see in the upper right corner the path to the directory where you have the downloaded data. Something like this:

.. image:: images/workshop-1/working-directory.png
   :alt: Spyder's Working Directory plugin showing the working directory downloaded-data-dir

Set up the virtual environment in Spyder
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Let's check that the virtual environment we created is enabled in Spyder. Go to Preferences > Python interpreter, and under "Use the following Python interpreter" add the path to your virtual environment. You should see something like this:

Reviewer 2: ```suggestion
Let's check that the virtual environment we created is enabled in Spyder. Go to Preferences > Python interpreter, and use the dropdown below "Use the following Python interpreter" to choose your virtual environment. You should see something like this:
```

.. image:: images/workshop-1/set-up-environment.png
   :alt: Set up environment in Spyder's Preferences

Now, you have everything ready to proceed with the workshop.
   
   
   
=====================
Obtain financial data
=====================
   

> When it comes to finance, being up to date is very important. So we are going to use a Python library that allows us to get updated historical Stock Market records from `Yahoo! Finance`_ API. In this way, we will be able to download data in the period of time we are interested in analyzing.

Reviewer 2: ```suggestion
When it comes to finance, being up to date is very important. So we are going to use a Python library that allows us to get updated historical Stock Market records from `Yahoo: Finance`_ API. In this way, we will be able to download data in the period of time we are interested in analyzing.
```

To get started, import the libraries.

.. important:: 

   Write and run all code for this workshop in the "IPython Console" at the bottom left of Spyder.

.. code-block:: python
      
   import numpy as np
   import pandas as pd
   from pylab import mpl
   from scipy.optimize import minimize
   import matplotlib.pyplot as plt
   import pprint
   
   from Historic_Crypto import Cryptocurrencies
   from Historic_Crypto import HistoricalData
   import yfinance as yf

The first group of imports is for basic operations. The second imports the libraries that we will use to download financial data.

Let's start exploring libraries with an example. We are going to get financial information about Netflix with one line of code:

.. code-block:: python

   netflix = yf.Ticker("NFLX")

We have used the ``Ticker`` class of the ``yfinance`` library to create a ``netflix`` object. This object contains attributes and methods that we can query to obtain various types of information.

General stock information
~~~~~~~~~~~~~~~~~~~~~~~~~

If you want to know which methods and attributes you can query, you can do so with the built-in ``help()`` function.

> You can also type the name of the object (``netflix``)in the console, then type a period and hit the Tab key once. IPython suggestions will then appear to help you navigate the object:

Reviewer 2: ```suggestion
You can also type the name of the object (``netflix``) in the console, then type a period and hit the Tab key once. IPython suggestions will then appear to help you navigate the object:
```

.. image:: images/workshop-1/ipython-auto-suggestions.png
   :alt: IPython Console auto-suggestions

For example, we can obtain general information with the ``info`` property of the object.

.. code-block:: python

   netflix_info = netflix.info

We can observe the result in the Variable Explorer, in the ``netflix_info`` variable. 

.. image:: images/workshop-1/net_flix-variable-in-variable-explorer.png
   :alt: First variable in Variable Explorer

If we double-click it, a new window with the detailed information will be displayed. 

.. image:: images/workshop-1/dict-in-variable-explorer.png
   :alt: Python dict example in Variable Explorer

> In that window we can see that a Python dictionary is displayed in which each key has a value assigned to it. Each type of value is represented by a distinctive color. For example, we see that Netflix is an Entertainment industry, the summary of the business type, among other indicators. The Variable Explorer is a very convenient way to view these types of results. We can compare it to viewing them in the IPython Console with the ``pprint`` function:

Reviewer 2: ```suggestion
In that window we can see a Python dictionary in which each key has a value assigned to it. Each type of value is represented by a distinctive color. For example, we see that Netflix is an Entertainment industry, the summary of the business type, among other indicators. The Variable Explorer is a very convenient way to view these types of results. We can compare it to viewing them in the IPython Console with the ``pprint`` function:
```

To measure the performance of our first portfolio we are going to use a classic theory in the world of finance: *mean-variance portfolio (MVP) theory*. This model assumes that investors only care about expected returns and the variance of such returns. The analysis is based entirely on statistical measures based on a time series of share prices, such as periodic mean returns and the variances of those returns with the same periodicity.

Prepare portfolio data
~~~~~~~~~~~~~~~~~~~~~~

Before we start, let's run some lines of code to style the plots.  These lines are optional, but we recommend that you run them so that the graphics look like the screenshots we present in this workshop.

.. code-block:: python

   plt.style.use("fivethirtyeight")
   mpl.rcParams["savefig.dpi"] = 300
   mpl.rcParams["font.family"] = "serif"
   np.set_printoptions(precision=5, suppress=True, formatter={"float": lambda x: f"{x:6.3f}"})

Suppose we want to measure the performance of a portfolio consisting of Google, Apple, Microsoft, Netflix and Amazon stocks. Let's call this set ``SYMBOLS_1``.

.. code-block:: python

   SYMBOLS_1 = ["GOOG", "AAPL", "MSFT", "NFLX", "AMZN"]

.. note::

>    If you search for ``SYMBOLS_1`` in Variable Explorer you will not find it: Python interprets this element not as a variable, but as a **constant**. This is because the name is written with uppercase letters (no letter is lowercase).

Reviewer 2: By default the Variable Explorer doesn't show this but it would be cool to mention that they can actually change the settings in the options menu to be able to see this variable.

> We are going to download the historical data for this portfolio. To do this, we are going to use the ``yfinance`` ``download()`` function, which takes as its first argument a string with the symbols (``SYMBOLS_1``) defined above. The rest of the arguments are the start date (``start="2012-01-01"``), the end date (``end="2021-01-01"``), and how the data will be grouped (``group_by="Ticker"``).

Reviewer 2: ```suggestion
We are going to download the historical data for this portfolio. To do this, we are going to use the ``yfinance`` ``download()`` function, which takes as its first argument a string with the symbols (``SYMBOLS_1``) defined above. The rest of the arguments are the start date (``start="2012-01-01"``), the end date (``end="2021-01-01"``) and how the data will be grouped (``group_by="Ticker"``).
```

If high return is desirable, high volatility is undesirable. The risk of this portfolio is relatively large.

Sharpe ratio
~~~~~~~~~~~~

The `Sharpe`_ ratio or index is a measure of portfolio performance. It relates the portfolio's return to its volatility, comparing the expected/realized return with the expected/realized risk. It is calculated as the difference between the actual investment returns and the expected return in a zero-risk situation, divided by the volatility of the investment. **It provides a model of the additional amount of returns received for each additional unit of risk**.

.. _Sharpe: https://en.wikipedia.org/wiki/Sharpe_ratio

Let's formalize this in a function:

.. code-block:: python

   def portfolio_sharpe(returns, weights):
       return portfolio_return(returns, weights) / portfolio_volatility(returns, weights)

And let's apply this to our portfolio.

.. code-block:: python

   portfolio_sharpe(rets_1, weights_1)

   # 1.2061518063427656

.. important::
   
   The Sharpe ratio measure is best understood in context: when comparing two or more portfolios, the one with the higher Sharpe ratio provides more profit for the same amount of risk.

> We can also use a Monte Carlo simulation to randomize the weights of each stock in the portfolio so that we can see the range over which the Sharpe ratio can vary. In this way we can plot some scenarios that together will give us a good insight into of the relationship between expected returns and expected volatility.

Reviewer 2: ```suggestion
We can also use a Monte Carlo simulation to randomize the weights of each stock in the portfolio so that we can see the range over which the Sharpe ratio can vary. In this way we can plot some scenarios that together will give us a good insight of the relationship between expected returns and expected volatility.
```

The mean expected return for our portfolio is 0.6972, which is higher than the realized return of 0.4181 that we would have obtained. Having invested in this portfolio over the long term (from 2016 to the present) would have been a very good deal. Due to the high variability, investing in the short term would have been very risky. In terms of gross profits, the realized returns of this portfolio were more than double those of portfolio 1 (0.4181 > 0.1997).

===========
Final words
===========

The mean-variance portfolio (MVP) theory is one of the many tools available to financial analysis. In recent years, machine learning algorithms have even been used to predict the behavior of stock prices more accurately than can be achieved with any standard financial theory.

The examples given during this workshop are not intended to serve as guidelines for you to invest your money. It is only a first step towards learning financial analysis using Python and a scientific IDE.

> With the skills learned here, you will be able to approach more complex topics of financial analysis such as those you will find in the references in the next section.

Reviewer 2: Maybe leave a"Homework" at the end such as in the previous workshop, maybe ask them to write the code for the third portfolio and compare their results with yours.

This seems to be a good portfolio because it has a good performance with a not very large variance.

Optimal portfolio weights
~~~~~~~~~~~~~~~~~~~~~~~~~

Can we use the data obtained to calculate the optimal weights for the portfolio by year? Of course we can. Let's start by delimiting the previous years as variables.

.. code-block:: python

   start_year, end_year = (2012, 2020)

We will now write a function to calculate these optimal weights.

.. code-block:: python

   def optimal_weights(returns, symbols, actual_weights, start_y, end_y):
      
       bounds = len(symbols) * [(0, 1), ]
       constraints = {"type": "eq", "fun": lambda weights: weights.sum() - 1}
       opt_weights = {}
 
       for year in range(start_y, end_y):
           _rets = returns[symbols].loc[f"{year}-01-01":f"{year}-12-31"]
           _opt_w = minimize(lambda weights: -portfolio_sharpe(_rets, weights), actual_weights, bounds=bounds, constraints=constraints)["x"]
           opt_weights[year] = _opt_w
       return opt_weights
       

> Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constraints`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constraints`` variables). Notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.

Reviewer: ```suggestion
Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constraints`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of SciPy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constraints`` variables). Notice the ``-`` sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of ``portfolio_sharpe`` a negative one.
```

.. code-block:: python

   port_1_exp_real[["expected", "realized"]].corr()

   #           expected  realized
   # expected  1.000000 -0.324053
   # realized -0.324053  1.000000

As we can see, the correlations are negative, which warns us that we should be cautious when using this type of modeling.

================
Second portfolio
================

We are now going to apply all the previous code with a portfolio of a different nature. Let's assume that instead of technology companies, we are now interested in pharmaceuticals. We will build a portfolio with stocks of Pfizer, Astra Zeneca, Johnson & Johnson.

Download the data
~~~~~~~~~~~~~~~~~

Let's download the data and format it.

.. code-block:: python

   SYMBOLS_2 = ["PFE", "AZN", "JNJ"]  # Pfizer, Astra Zeneca, Johnson N Johnson

   data_2 = yf.download(" ".join(SYMBOLS_2), start="2012-01-01", end="2021-01-01", group_by="Ticker")
   data_2 = data_2.stack(level=1).rename_axis(["Date", "Ticker"]).reset_index(level=1)

   close_data_2 = data_2[data_2.Ticker == "Close"].drop("Ticker", inplace=False, axis=1)

.. note::

>    If you do not want to use the yfinance API, you can download the close_data_2.csv file containing the closing information for this portfolio. Copy this file to your working directory. Load the data with the following instruction:  ``>>> close_data_2 = pd.read_csv("close_data_2.csv")``.

Reviewer: ```suggestion
   If you do not want to use the yfinance API, you can download the ``close_data_2.csv`` file containing the closing information for this portfolio. Copy this file to your working directory. Load the data with the following instruction:  ``>>> close_data_2 = pd.read_csv("close_data_2.csv")``.
```

.. _The Python Tutorial: https://docs.python.org/3/tutorial/

You will also need to have Spyder installed. More information about installation in :doc:`installation guide<../installation>`.

.. important::

   Spyder now offers :ref:`standalone_installers_ref` for Windows and macOS, making it easier to get up and running with the application without having to download Anaconda or manually install it in your existing environment.
   While we still support Anaconda, we recommend this install method on those platforms to avoid most problems with package conflicts and other issues.

==============
Learning goals
==============

After completing this workshop, you should be able to:

* Apply elementary statistical analysis to stock and cryptocurrency portfolios to measure their performance
* Understand the advantages of programming with an IDE

===============
Learner profile
===============

This workshop is intended for people interested in finance who want to take their first steps in Financial Analysis using Python and Spyder.

===============

> Prior knowledge

Reviewer 2: I think you missed this one

=====
Intro
=====

In this workshop we will obtain financial data in real-time and explore financial portfolios using Econometrics and computational tools.

Why Python for financial analysis?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Real-time analysis of historical and current financial data is essential for those investing in financial instruments. Real-time analysis of historical and current financial data is essential for those investing in financial instruments. Python has a number of features that make it ideal for financial tasks:

* It is easy to learn for anyone, whether they have previous programming experience or not
* Python has a variety of specialized mathematical and statistical libraries (SciPy, NumPy, Pandas) that are commonly used for financial analysis
* It can be connected to APIs for loading financial data in real time
* It is the programming language with the most resources for machine learning (supervised learning, unsupervised learning, reinforcement learning), which is one of the most useful tools for Econometrics nowadays
* It has excellent libraries for plotting
* You can use resources such as Google Colab or Binder to do your analysis in the cloud.

> Why should I use an IDE?

Reviewer 2: Missed this one too

~~~~~~~~~~~~~~~~~~~~~~~~

Although you can use Python without an IDE (Integrated Development Environment), you will work much better with one. Spyder is a Scientific Integrated Development Environment written in Python, and designed by and for scientists, engineers, and data analysts. It features a unique combination of the advanced editing, analysis, debugging, and profiling functionality of a comprehensive development tool with the data exploration, interactive execution, deep inspection, and beautiful visualization capabilities of a scientific package. These capabilities and its integration with Python make Spyder work perfectly for financial analysis.

Introduction to financial analysis with Spyder
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you're not familiar with Spyder, we recommend you start with our :doc:`Quickstart<../quickstart>`. But if you want a  summary, here's a quick overview.

Reviewer 2: I think this section should be much shorter If we are leaving it

.. image:: images/workshop-1/ipython-auto-suggestions.png
   :alt: IPython Console auto-suggestions

For example, we can obtain general information with the ``info`` property of the object.

.. code-block:: python

   netflix_info = netflix.info

You can observe the result in the Variable Explorer, in the ``netflix_info`` variable. 

.. image:: images/workshop-1/net_flix-variable-in-variable-explorer.png
   :alt: First variable in Variable Explorer

If you double-click on it, a new window with the detailed information will be displayed. 

.. image:: images/workshop-1/dict-in-variable-explorer.png
   :alt: Python dict example in Variable Explorer

In that window you will see that a Python dictionary is displayed in which each key has a value assigned to it. Each type of value is represented by a distinctive color. For example, you can see that Netflix is an Entertainment industry, the summary of the business type, among other indicators. The Variable Explorer is a very convenient way to view these types of results. You can compare it to viewing them in the Ipython Console with the ``pprint`` function:

.. code-block:: python

   pprint.pprint(netflix_info)

> Although ``pprint`` displays all the information, it is easier to view it in the Variable Explorer.

Reviewer 2: Missed this one

   # 1.2061518063427656

.. important::
   
   The Sharpe ratio measure is best understood in context: when comparing two or more portfolios, the one with the higher Sharpe ratio provides more profit for the same amount of risk.

We can also use a Monte Carlo simulation to randomize the weights of each stock in the portfolio so that we can see the range over which the Sharpe ratio can vary. In this way we can plot some scenarios that together will give us a good insight into of the relationship between expected returns and expected volatility.

We are going to do this with a function that we will explain step by step.

.. code-block:: python

   def monte_carlo_sharpe(returns, symbols, weights):

       sim_weights = np.random.random((1000, len(symbols)))
       sim_weights = (sim_weights.T / sim_weights.sum(axis=1)).T

       volat_ret = [(portfolio_volatility(returns[symbols], weights),
                   portfolio_return(returns[symbols], weights))
                   for weights in sim_weights]
       volat_ret = np.array(volat_ret)

       sharpe_ratio = volat_ret[:, 1] / volat_ret[:, 0]

       return volat_ret, sharpe_ratio

.. warning::

>    You do not need to type the following code in the IPython Console. If you write the function above it will be enough. It is just a code presentation to explain what is inside the function.

Reviewer 2: You missed this one

>    You do not need to type the following code in the IPython Console. If you write the function above it will be enough. It is just a code presentation to explain what is inside the function.

Author: But if I do that, doesn't it seem to you that people might write the lines of code separately in the console, causing more confusion?

The code lines:

.. code-block:: python

   sim_weights = np.random.random((1000, len(symbols)))
   sim_weights = (sim_weights.T / sim_weights.sum(axis=1)).T

create a numpy array of length 1,000 and width of the number of shares in the portfolio. Each row of the array has random weights that always add up to 1.

The next section:

.. code-block:: python

   volat_ret = [(portfolio_volatility(returns[symbols], weights),
                   portfolio_return(returns[symbols], weights))
                   for weights in sim_weights]
   volat_ret = np.array(volat_ret)

calculates the volatility and returns for the new random weights using a list comprehension. The resulting list is transformed back into a numpy array.

> Finally:

Reviewer: I agree with @juanis2112 , this part is a bit confusing and could be reworded.

This seems to be a good portfolio because it has a good performance with a not very large variance.

Optimal portfolio weights
~~~~~~~~~~~~~~~~~~~~~~~~~

Can we use the data obtained to calculate the optimal weights for the portfolio by year? Of course we can. Let's start by delimiting the previous years as variables.

.. code-block:: python

   start_year, end_year = (2012, 2020)

We will now write a function to calculate these optimal weights.

.. code-block:: python

   def optimal_weights(returns, symbols, actual_weights, start_y, end_y):
      
       bounds = len(symbols) * [(0, 1), ]
       constrains = {"type": "eq", "fun": lambda weights: weights.sum() - 1}
       opt_weights = {}
 
       for year in range(start_y, end_y):
           _rets = returns[symbols].loc[f"{year}-01-01":f"{year}-12-31"]
           _opt_w = minimize(lambda weights: -portfolio_sharpe(_rets, weights),
                          actual_weights,
                          bounds=bounds,
                          constraints=constrains)["x"]
           opt_weights[year] = _opt_w
       return opt_weights
       

> Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy (which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, and the ``bounds`` and the ``constrains`` variables). Notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.

Reviewer 2: ```suggestion
Let's describe this function in broad strokes. ``bounds`` indicates the maximum and minimum weights for each stock in the portfolio. The lowest weight will be 0 and the highest weight will be 1 for each stock in the portfolio. ``constrains`` is a function that ensures that the sum of the weights of all actions always adds up to 1. Then a loop is initialized that will segment the data for each year. In the variable ``_rets`` the returns for the specified year are obtained. In ``_opt_w`` the ``portfolio_shape()`` function is used to calculate the weights that maximize the Sharpe ratio. This is done with the ``minimize()`` function of Scipy, which takes as arguments the ``portfolio_shape`` function, the actual weights of our stocks in the portfolio, the ``bounds`` and the ``constrains`` variables. Did you notice the - sign before ``portfolio_sharpe``? It's because ``minimize()`` aims to find the minimum value of a function relative to a parameter, but we are interested in the maximum, so we make the result of portfolio_sharpe a negative one.
```

We will use the function we just defined to calculate the optimal weights for each year. And we are going to save the result in a Pandas DataFrame to take advantage of the Variable Explorer display options.

.. code-block:: python

   opt_weights_1 = optimal_weights(rets_1, SYMBOLS_1, weights_1, start_year, end_year)
   port_1_ow = pd.DataFrame.from_dict(opt_weights_1, orient='index')
   port_1_ow.columns = SYMBOLS_1

> Double click on the ``port_1_ow`` variable in the Variable Explorer. In the table, you can specify the number of decimal places to display, as you can see in the following image:

Reviewer 2: ```suggestion
Double click on the ``port_1_ow`` variable in the Variable Explorer. In the table, you can specify the number of decimal places to display, by clicking the format button at the bottom left of the pane, as you can see in the following image:
```

.. image:: images/workshop-1/portfolio-1-optimal-weights.png
   :alt: Portfolio 1 optimal weights by year

> We set the number of decimal places to 4 and unchecked the **Column min/max** checkbox to better appreciate the contrasts in the row values (years).

Reviewer 2: ```suggestion
We set the number of decimal places to 4 and uncheck the **Column min/max** checkbox to better appreciate the contrasts in the row values (years).
```
